src/category_theory/limits/cofinal.lean

```diff
@@ -0,0 +1,225 @@
+    naturality' := λ X Y f,
+    begin
+      dsimp, simp,
+      -- This would be true if we'd chosen `lift F X` to be `lift F Y`
+      -- and `hom_to_lift F X` to be `f ≫ hom_to_lift F Y`.
+      apply induction F (λ Z k, G.map f ≫ G.map (hom_to_lift F Y) ≫ c.ι.app (lift F Y) = G.map k ≫ c.ι.app Z),
+      { intros Z₁ Z₂ k₁ k₂ g a z,
+       rw [←a, functor.map_comp, category.assoc, ←functor.comp_map, c.w, z], },
+      { intros Z₁ Z₂ k₁ k₂ g a z,
+       rw [←a, functor.map_comp, category.assoc, ←functor.comp_map, c.w] at z,
+       rw z, },
+      { rw [←functor.map_comp_assoc], },
+    end }}
+
+variables (F)
+
+@[simp]
+lemma colimit_cocone_comp_aux (s : cocone (F ⋙ G)) (j : C) :
+  G.map (hom_to_lift F (F.obj j)) ≫ s.ι.app (lift F (F.obj j)) =
+    s.ι.app j :=
+begin
+  -- This point is that this would be true if we took `lift (F.obj j)` to just be `j`
+  -- and `hom_to_lift (F.obj j)` to be `𝟙 (F.obj j)`.
+  apply induction F (λ X k, G.map k ≫ s.ι.app X = (s.ι.app j : _)),
+  { intros j₁ j₂ k₁ k₂ f w h, rw ←w, rw ← s.w f at h, simpa using h, },
+  { intros j₁ j₂ k₁ k₂ f w h, rw ←w at h, rw ← s.w f, simpa using h, },
+  { exact s.w (𝟙 _), },
+end
+
+/-- Given a colimit cocone over `G` we can construct a colimit cocone over `F ⋙ G`. -/
+@[simps]
+def colimit_cocone_comp (t : colimit_cocone G) :
+  colimit_cocone (F ⋙ G) :=
+{ cocone := t.cocone.whisker F,
+  is_colimit :=
+  { desc := λ s, t.is_colimit.desc (extend_cocone s),
+    uniq' := λ s m w,
+    begin
+      apply t.is_colimit.hom_ext,
+      intro d,
+      simp [(w _).symm],
+    end, }, }.
+
+@[priority 100]
+instance comp_has_colimit [has_colimit G] :
+  has_colimit (F ⋙ G) :=
+has_colimit.mk (colimit_cocone_comp F (get_colimit_cocone G))
+
+lemma colimit_pre_is_iso_aux (t : colimit_cocone G) :
+  (colimit_cocone_comp F t).is_colimit.desc (t.cocone.whisker F) = 𝟙 t.cocone.X :=
+begin
+  apply t.is_colimit.hom_ext,
+  tidy,
+end
+
+instance colimit_pre_is_iso [has_colimit G] :
+  is_iso (colimit.pre G F) :=
+begin
+  rw colimit.pre_eq (colimit_cocone_comp F (get_colimit_cocone G)) (get_colimit_cocone G),
+  rw colimit_pre_is_iso_aux,
+  dsimp,
+  apply_instance,
+end
+
+/--
+When `F` is cofinal, and `G` has a colimit, then `F ⋙ G` has a colimit also and
+`colimit (F ⋙ G) ≅ colimit G`
+
+https://stacks.math.columbia.edu/tag/04E7
+-/
+def colimit_iso [has_colimit G] : colimit (F ⋙ G) ≅ colimit G := as_iso (colimit.pre G F)
+
+/-- Given a colimit cocone over `F ⋙ G` we can construct a colimit cocone over `G`. -/
+@[simps]
+def colimit_cocone_of_comp (t : colimit_cocone (F ⋙ G)) :
+  colimit_cocone G :=
+{ cocone := extend_cocone t.cocone,
+  is_colimit :=
+  { desc := λ s, t.is_colimit.desc (s.whisker F),
+    uniq' := λ s m w,
+    begin
+      apply t.is_colimit.hom_ext,
+      intro X,
+      simp [(w _).symm],
+    end, }, }
+
+/--
+When `F` is cofinal, and `F ⋙ G` has a colimit, then `G` has a colimit also.
+
+We can't make this an instance.
```

Comment:
I presume this is because it clashes with `comp_has_colimit`, but could you add this to the docstring for future reference?

Reply:
Done!